### PR TITLE
Add Building a GitHub Secrets Scanner

### PR DESCRIPTION
**Pull request by jpf**
This blog post has the following features:

- [X] A title that's approved by Dev Advocacy
- [X] A URL that's approved by Dev Advocacy
- [X] Rendered locally and confirmed that no Markdown typos exist

### CODE REVIEW

**Comment by mraible** [comment]
It seems strange to have contractors writing awareness posts. At first glance, it seems like a lot of line breaks are needed between sections (for readability mostly, markdown might handle it OK).

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+layout: blog_post
+title: Building a GitHub Secrets Scanner
+author: vickie-li
+by: contractor
+communities: [python]
+description: "Use Python to build a basic GitHub secrets scanner."
+tags: [python]
+tweets:
+- "Learn the basics of how to scan GitHub for sensitive information."
+- "Would you know if someone checked secrets into GitHub? Learn how to build a tool to check."
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:11]: Can you add another Tweet, please? Our "from the archives" script expects 3.

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+image: blog/building-a-github-secrets-scanner/card.png
+type: awareness
+---
+GitHub reconnaissance is a tactic that attackers use to gather information about their targets. Attackers analyze organizations' GitHub repositories and check for sensitive data that has been accidentally committed or information that could lead to the discovery of a vulnerability.
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:15]:
```suggestion

GitHub reconnaissance is a tactic that attackers use to gather information about their targets. Attackers analyze organizations' GitHub repositories and check for sensitive data that has been accidentally committed or information that could lead to the discovery of a vulnerability.
```

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+For this tutorial, let's build a scanner that automates the GitHub recon process! We will be scanning your GitHub repositories using the method mentioned in the "Tightening Up Your GitHub Security" post.
+We will be working with Python, the GitHub REST API, and the GitPython Library. You will need to obtain a [GitHub personal access token](https://github.com/settings/tokens) and [install GitPython](https://gitpython.readthedocs.io/en/stable/intro.html#installing-gitpython).
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:17]:
```suggestion

You will be working with Python, the GitHub REST API, and the GitPython Library. You will need to obtain a [GitHub personal access token](https://github.com/settings/tokens) and [install GitPython](https://gitpython.readthedocs.io/en/stable/intro.html#installing-gitpython). 
```

I could also see these items listed in a prerequisites bullet-list like we do in other posts.

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+- "Would you know if someone checked secrets into GitHub? Learn how to build a tool to check."
+image: blog/building-a-github-secrets-scanner/card.png
+type: awareness
+---
+GitHub reconnaissance is a tactic that attackers use to gather information about their targets. Attackers analyze organizations' GitHub repositories and check for sensitive data that has been accidentally committed or information that could lead to the discovery of a vulnerability.
+For this tutorial, let's build a scanner that automates the GitHub recon process! We will be scanning your GitHub repositories using the method mentioned in the "Tightening Up Your GitHub Security" post.
+We will be working with Python, the GitHub REST API, and the GitPython Library. You will need to obtain a [GitHub personal access token](https://github.com/settings/tokens) and [install GitPython](https://gitpython.readthedocs.io/en/stable/intro.html#installing-gitpython). 
+Make sure that your GitHub personal access token has these permissions: `public_repo`, `read:org`, `read:packages`, `read:user`, `repo:status`
+Before we dive into the code, here are the imports you need:
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:19]:
```suggestion
Before you dive into the code, here are the imports you need:

```

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+```python
+import requests, tempfile, shutil, re
+
+from git import Repo
+from git import NULL_TREE
+```
+## Finding Repositories
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:26]:
```suggestion

## Finding Repositories
```

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+First, we will crawl through your organization's account to find repositories to analyze. We need the GitHub API for this. The GitHub API endpoint is located at `https://api.github.com`.
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:27]:
```suggestion
First, you will crawl through your organization's account to find repositories to analyze. You need the GitHub API for this. The GitHub API endpoint is located at `https://api.github.com`.
```

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+You can set the "Accept" header to "application/vnd.github.v3+json" to specify that we are using the REST API v3. And you can provide the access token through an "Authorization" header.
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:28]:
```suggestion

You can set the "Accept" header to "application/vnd.github.v3+json" to specify that you are using the REST API v3. And you can provide the access token through an "Authorization" header.

```

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+headers = {
+    "Authorization": "token {}".format(GITHUB_ACCESS_TOKEN),
+    "Accept": "application/vnd.github.v3+json",
+}
+api_host = "https://api.github.com"
+
+orgs = ["okta"]  # list organizations to analyze
+users = ["vickie"]  # list users to analyze
+```
+List the organization and individual usernames that you want to analyze. Include the usernames of your organization as well as the usernames of your organization's employees. Then, we'll start by looking for additional usernames associated with your organization by querying the GitHub API for organization members.
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:41]:
```suggestion

List the organization and individual usernames that you want to analyze. Include the usernames of your organization as well as the usernames of your organization's employees. Then, you'll start by looking for additional usernames associated with your organization by querying the GitHub API for organization members.

```

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+```python
+def requests_page(path):
+    resp = requests.Response()
+    try:
+        resp = requests.get(url=path, headers=headers, timeout=15)
+    except:
+        pass
+    return resp.json()
+
+
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:51]: Remove extra line break.

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+def find_members():
+    """Find usernames associated with the organizations."""
+    resp = []
+    for org in orgs:
+        path = "{}/orgs/{}/members".format(api_host, org)
+        resp = requests_page(path)
+    for user in resp:
+        users.append(user["login"])
+```
+Optionally, you can look for contributors to your organization's projects. You can also update your `users` and `orgs` list manually as you go.
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:61]:
```suggestion

Optionally, you can look for contributors to your organization's projects. You can also update your `users` and `orgs` list manually as you go.

```

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+```python
+def find_contributors(org, repo):
+    """Find the contributors of found repos and add to users list."""
+    path = "{}/repos/{}/{}/collaborators".format(api_host, org, repo)
+    resp = requests_page(path)
+    for user in resp:
+        users.append(user["login"])
+```
+Next, we scan the usernames for repositories and clone the repo into a temporary directory for analysis. You might have to figure out some permission issues with the GitHub API during this step. Make sure that your access tokens have the proper permissions!
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:70]:
```suggestion

Next, you scan the usernames for repositories and clone the repo into a temporary directory for analysis. You might have to figure out some permission issues with the GitHub API during this step. Make sure that your access tokens have the proper permissions!

```

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+            scan_repo(repo["git_url"])
+
+
+def clone_repo(repo_url):
+    """Create a temporary directory to hold the repo that we are analyzing."""
+    temp_dir = tempfile.mkdtemp()
+    Repo.clone_from(repo_url, temp_dir)
+    return temp_dir
+```
+## Building a Wordlist
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:90]:
```suggestion

## Building a Wordlist

```

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+You can also build a list of keywords that might indicate vulnerabilities or an information leak. Add search terms based on your needs. You can incorporate regex patterns as well to scan for access tokens or keys with a known format.
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:91]:
```suggestion
You can also build a list of keywords that might indicate vulnerabilities or an information leak. Add search terms based on your needs. You can incorporate regex patterns as well to scan for access tokens or keys with a known format.

```

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+        "dev",
+        "removed",
+        "HTTP",
+        "HTTPS",
+        "FTP",
+    ],
+    "weak crypto": ["ECB", "MD4", "MD5"],
+}
+```
+## Scanning Repositories
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:154]:
```suggestion

## Scanning Repositories

```

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+Now, let's start scanning our repositories for keywords! We can use the GitPython library to iterate through the commits of each repository and scan for the strings in our dictionary. Finally, we display the filenames and commit numbers that might be leaking information.
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:155]:
```suggestion
Now, let's start scanning your repositories for keywords! You can use the GitPython library to iterate through the commits of each repository and scan for the strings in your dictionary. Finally, it displays the filenames and commit numbers that might be leaking information.

```

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+```python
+class colors:
+    FILENAME = "\033[1m"
+    NORMAL = "\033[0m"
+    WARNING = "\033[93m"
+
+
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:162]: Remove blank line.

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+                        colors.WARNING,
+                        term,
+                        colors.NORMAL,
+                        colors.WARNING,
+                        category,
+                        colors.NORMAL,
+                    )
+                )
+
+
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:180]: Remove blank line.

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+                )
+            )
+    diffs = commit.diff(prev_commit, create_patch=True)
+    for diff in diffs:
+        find_strings(diff)
+        prev_commit = commit
+    # remember to remove the tempfile
+    shutil.rmtree(temp_dir)
+
+
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:203]: Remove blank line.

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+if __name__ == "__main__":
+    find_members()
+    find_repos()
+```
+## Eliminating GitHub Security Holes
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:208]:
```suggestion

## Eliminating GitHub Security Holes

```

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 20]
```diff
@@ -0,0 +1,219 @@
...
+In this post, we built a simple GitHub scanner that finds secrets accidentally committed to your organization's repositories. If this is something that interests you, here are some related project ideas.
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:209]:
```suggestion
In this post, you built a simple GitHub scanner that finds secrets accidentally committed to your organization's repositories. If this is something that interests you, here are some related project ideas.
```

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+After you've identified your GitHub weaknesses, here are a few steps that you can take to tighten up your GitHub security. First, every piece of sensitive data exposed through a public GitHub repository should be considered leaked. Therefore, you should rotate every credential found in your repositories. Next, fix or patch vulnerabilities that you have found via code analysis of these files. Again, these weaknesses should be considered public information and you should remediate them ASAP. Finally, [remove any additional sensitive information](https://help.github.com/en/github/authenticating-to-github/removing-sensitive-data-from-a-repository) from the repositories and commit histories!
```
Comment (mraible) [inline _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:211]:
```suggestion

After you've identified your GitHub weaknesses, here are a few steps that you can take to tighten up your GitHub security. First, every piece of sensitive data exposed through a public GitHub repository should be considered leaked. Therefore, you should rotate every credential found in your repositories. Next, fix or patch vulnerabilities that you have found via code analysis of these files. Again, these weaknesses should be considered public information and you should remediate them ASAP. Finally, [remove any additional sensitive information](https://help.github.com/en/github/authenticating-to-github/removing-sensitive-data-from-a-repository) from the repositories and commit histories!
```

**Review by mraible — APPROVED** [review]

File: _source/_posts/2021-02-01-building-a-github-secrets-scanner.md (hunk 9 above)
Comment (jpf) [reply _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:60]: The extra blank lines are part of standard Python formatting using the `black` idiomatic source code formatting tool.

In this case, the extra line break is part of the PEP-8 standard: https://www.python.org/dev/peps/pep-0008/#blank-lines

> Surround top-level function and class definitions with two blank lines.

Comment (mraible) [reply _source/_posts/2021-02-01-building-a-github-secrets-scanner.md:60]: Interesting. I did not know that. Thanks for the info!

**Pull request merged**